### ISSUE TITLE
Integration tests should sometimes cleanly shut nodes down

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#35585]
Today in our real-node integration tests we shut nodes down with `kill -9 $PID`, i.e. sending `SIGKILL` to the process, which shuts it down immediately. Although it is very important to test the behaviour of nodes shutting down like this, I think the fact that we always use `SIGKILL` means we're failing to properly cover the normal shutdown procedure that happens in response to a less aggressive signal.

I think we should also send `SIGTERM` in these tests.

**Comment by atorok** [comment]
Test clusters will terminate with `TERM` by default.
( I just wrote the code for it, will have a PR shortly ).

**Comment by atorok** [comment]
@DaveCTurner do you think we should send `TERM` on a list of specific tests or just implement some tests specifically for this ?

**Comment by DaveCTurner** [comment]
Do we have access to deterministic randomness at the point we're making the choice? I think all tests should work correctly whether the nodes are stopped with `TERM` or `KILL` (or a mixture) so the best would be to choose randomly on each test, weighting `KILL` more heavily since we prefer that for speed reasons.

**Comment by jasontedor** [comment]
I would prefer that we not introduce that kind of randomness into the build. We use to have that kind of randomness all over the place in the build (which GC to use, which JVM flags to use, etc.) and it made debugging a nightmare.

What is the plan with "testing" that we work correctly with `SIGTERM`? We don't want the build to wait forever, but if we introduce a timeout, then we are adding another vector for flaky tests where we end up failing the build because the process didn't shutdown quickly enough after receiving `SIGTERM`. Are there post-assertions that we would be making, or only that the process eventually shutdown?

**Comment by DaveCTurner** [comment]
The thing I think we should be checking is that node(s) that were shut down with `SIGTERM` come up and join the cluster again: particularly `full-cluster-restart` and `rolling-upgrade`. I think we don't currently have any that restart a cluster without upgrading, but if we did then the same thing too.

Sorry, this was not clear from the OP. We don't really care about the behaviour of a node that shuts down never to start again, so `SIGKILL` is fine for them.

I don't think failing to shut down within a timeout should be considered a failure; in fact I think we should also be checking that a `SIGTERM` followed by a `SIGKILL` results in correct behaviour on a restart.

**Comment by atorok** [comment]
@DaveCTurner all the bwc tests have a version where they from current version to  current  version with no actual upgrade in-between. 

It would be nice to have focused tests. Correct me if I'm wrong but it would seem that in this case the upgrade might be relevant, but the tests ran against the nodes are not really relevant as we are just looking to make sure the cluster is formed.  Full cluster restart vs rolling doesn't seem particularly relevant either. 

Thus we could have a bwc test similar to full cluster restart that kills nodes with `TERM` instead of `KILL` and runs a single test to make sure the cluster is formed.

**Comment by DaveCTurner** [comment]
Ah I'm pleased to hear that we're running `CURRENT` -> `CURRENT` non-upgrade upgrades, thanks for that.

I'm on the fence here. I think we want to be checking a lot more than just the cluster is formed. We run quite a lot of code in reaction to a `SIGTERM`, some of which makes changes on disk. I think we shouldn't be skipping over this in all real-node tests. But on the other hand we are shutting nodes down cleanly in `EsIntegTestCase`-based tests so maybe the gap isn't so big.

**Comment by rjernst** [comment]
> Thus we could have a bwc test similar to full cluster restart that kills nodes with TERM instead of KILL and runs a single test to make sure the cluster is formed.

We could do this similar to how we alter the distribution tested for qa projects, by having a setting on the testcluster that changes how the node is shutdown, and changing that in a dedicated CI job for the bwc tests. I would be happy with this if we only tested CURRENT -> CURRENT for the alternate shutdown method, so the dedicated CI job should run fast.

**Comment by droberts195** [comment]
> in fact I think we should also be checking that a SIGTERM followed by a SIGKILL results in correct behaviour on a restart.

Yes, definitely, because this is exactly how nodes are shut down in cloud: first a SIGTERM is sent, but after a timeout a SIGKILL is sent if ES is still running.  I believe the timeout has been changed over the years.  Ideally the tests should use the same timeout that cloud is using or a shorter one to catch the cases where killing part way through graceful termination causes a problem that would affect cloud users if not found in testing.
